jax/_src/interpreters/mlir.py

```diff
@@ -1383,6 +1383,25 @@ def update_tokens(self, tokens: TokenSet) -> TokenSet:
         new_tokens.append((eff, self._tokens[eff]))
     return TokenSet(new_tokens)
 
+def check_for_large_number_of_constants(jaxpr):
+  if (thresh := config.captured_constants_warn_threshold.value) < 0:
+    return
+
+  consts_size = sum(map(operator.attrgetter("size"), jaxpr.consts))
+  # staging this here avoids division by zero and unnecessary querying of in_avals size
+  if consts_size == 0:
+    return
+
+  in_avals_size = sum(map(operator.attrgetter("size"), jaxpr.in_avals))
+  if consts_size > (consts_size + in_avals_size) * thresh:
+    message = (
+      "Large number of captured constants detected during lowering "
```

Comment:
Why not keep it disabled first, try it out with users and then once there is enough evidence that doing this is indeed bad, you can turn it on to 25%?

Issuing warnings by default can get annoying and asking users to opt-out is not a good experience.

Reply:
Experiencing long compile times is also annoying. Provided the false negative rate is low I think we should issue a warning by default. But 0.25 seems like it's too low for a large portion of the test suite.

Reply:
> Experiencing long compile times is also annoying

It can be due to any number of reasons (not using scan for example). I don't think the rate is low. You have no evidence for that. And test suite is not the right tool for this measurement since we don't create large arrays in test suite.

Reply:
I believe you have misunderstood me. I did not claim that the false positive rate is low or high, or that the the test suite is representative. It probably is not.

Reply:
My higher level point was that make this opt-out instead of opt-in. Currently, it's opt-in IIUC?

Reply:
Sorry, the opposite way lol.

My higher level point was that make this opt-in for users instead of opt-out. Currently it's opt-out for users IIUC?

Reply:
Correct. Regarding your high level point, I think the inconvenience of capturing constants and attributing unnecessarily inflated compile times to JAX being slow is a greater problem than potentially incurring an errant warning. 

This seems exactly like the instance when you would like to provide the the user with a heads up that they might be doing something bad by mistake.

Reply:
> slow is a greater problem than potentially incurring an errant warning.

I disagree. But I'll let @mattjj respond. I think I have made my point.

The same argument can be made for raising warnings by default for every thing that can lead to slow compile times and that can get annoying if it's opt-out. 

But if it's opt-in, then it sounds like a good idea. I disagree on the opt-opt part and it's not just an errant warning.